### ISSUE TITLE
responses with external $ref thrown nil pointer dereference

### ISSUE THREAD

**Issue by BertrandGouny** [go-swagger/go-swagger#1406]
## Problem statement
Using an external $ref definition for responses thrown nil pointer dereference:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x8bf5f0]

goroutine 1 [running]:
github.com/go-swagger/go-swagger/vendor/github.com/go-openapi/validate.(*SpecValidator).validateResponseExample(0xc421d27c80, 0xc4202b34c8, 0x1, 0xc421d27818, 0xc42008c780)
	/home/hello/workspace/project/go/src/github.com/go-swagger/go-swagger/vendor/github.com/go-openapi/validate/spec.go:837 +0x3d0
github.com/go-swagger/go-swagger/vendor/github.com/go-openapi/validate.(*SpecValidator).validateExamplesValidAgainstSchema(0xc421d27c80, 0xc421d27a90)
	/home/hello/workspace/project/go/src/github.com/go-swagger/go-swagger/vendor/github.com/go-openapi/validate/spec.go:876 +0x281
github.com/go-swagger/go-swagger/vendor/github.com/go-openapi/validate.(*SpecValidator).Validate(0xc421d27c80, 0xb610e0, 0xc4202a4690, 0x0, 0x0)
	/home/hello/workspace/project/go/src/github.com/go-swagger/go-swagger/vendor/github.com/go-openapi/validate/spec.go:161 +0x4bb
github.com/go-swagger/go-swagger/cmd/swagger/commands.(*ValidateSpec).Execute(0xc4202b2608, 0xc4200e7d00, 0x1, 0x2, 0xc4202b2608, 0x1)
	/home/hello/workspace/project/go/src/github.com/go-swagger/go-swagger/cmd/swagger/commands/validate.go:60 +0x168
github.com/go-swagger/go-swagger/vendor/github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc420159320, 0xc420010190, 0x2, 0x2, 0x412608, 0xc420490630, 0xc42018fe00, 0xc4200e7201, 0xc420257d40)
	/home/hello/workspace/project/go/src/github.com/go-swagger/go-swagger/vendor/github.com/jessevdk/go-flags/parser.go:316 +0x841
github.com/go-swagger/go-swagger/vendor/github.com/jessevdk/go-flags.(*Parser).Parse(0xc420159320, 0x6, 0xbb0ebe, 0x6, 0x0, 0xbf88d9)
	/home/hello/workspace/project/go/src/github.com/go-swagger/go-swagger/vendor/github.com/jessevdk/go-flags/parser.go:186 +0x71
main.main()
	/home/hello/workspace/project/go/src/github.com/go-swagger/go-swagger/cmd/swagger/swagger.go:140 +0xd89
```

## Swagger specification

responses.yaml: 
```
swagger: '2.0'
info:
  title: Responses
  version: 0.1.0
responses:
  NotFound:
    description: Not found

paths:
  /:
    get:
      summary: GET
      operationId: getAll
      responses:
        '200':
          description: Ok
```

swagger.yaml: 
```
swagger: '2.0'
info:
  title: Object
  version: 0.1.0

paths:
  /:
    get:
      summary: GET
      operationId: getAll
      responses:
        '200':
          description: Ok
        '404':
          $ref: './responses.yaml#/responses/NotFound'

```

## Steps to reproduce
```
swagger validate swagger.yaml
```
## Environment
swagger version: tested with 0.13.0 and dev
go version: 1.9.2
OS: Ubuntu 16.04

**Comment by fredbi** [comment]
This construct is not legal in Swagger 2.0. $ref should be under the schema section.
The panic issue has been fixed by go-openapi/validate#72 and should be soon vendored with go-swagger

**Comment by BertrandGouny** [comment]
Thanks for the quick reply :) 

not sure if this is not legal, here an example from the swagger 2 documentation (maybe outdated ?) : 
https://swagger.io/docs/specification/2-0/describing-responses/ 
```
        401:
          $ref: "#/responses/Unauthorized"   # <-----
```

**Comment by fredbi** [comment]
You're right. I have been misled by the bug I corrected yesterday with invalid $ref in response.
I have to check your case with the newest go-openapi/validate before releasing it to go-swagger.
Thanks

**Comment by gregmarr** [comment]
This is explicit in the spec too, it lists Response Object or Reference Object as the content for a status code.
Any HTTP status code can be used as the property name (one property per HTTP status code). Describes the expected response for that HTTP status code. Reference Object can be used to link to a response that is defined at the Swagger Object's responses section.

**Comment by fredbi** [comment]
Yeah I found it. Strange that none of the 00's fixtures used to test validate did no exhibit this pattern.
Have to check how the spec package renders to $ref to validate. It is supposed to expand it as a Response object. I'll look at it tonight.

**Comment by fredbi** [comment]
OK I tested it more seriously. Here is my takeaway: there are some good and bad news.
- I could fix the panic issue. It was actually a test case I was missing.
- The #/responses does not resolve to external references (same for #/parameters)

   It looks like $ref to external files are only resolved for schema objects.

   That should be confirmed by our $ref-resolver experts around (@kreamyx, @casualjim)

**Comment by kreamyx** [comment]
@fredbi it seems like `swagger flatten` is working correctly
`go run cm/swagger/swagger.go responses.yaml`
```json
{
  "swagger": "2.0",
  "info": {
    "title": "Object",
    "version": "0.1.0"
  },
  "paths": {
    "/": {
      "get": {
        "summary": "GET",
        "operationId": "getAll",
        "responses": {
          "200": {
            "description": "Ok"
          },
          "404": {
            "description": "Not found"
          }
        }
      }
    }
  }
}
```

**Comment by kreamyx** [comment]
oh I see the issue has been closed. Sorry for the late reply :)

**Comment by fredbi** [comment]
@kreamyx  not really closed. I opened a side issue on validate regarding the panic behavior, which is now closed.
But I needed your advice on the $ref resolution part (which is in spec, not validate).
I added some fixtures in validate to exhibit this behavior:
- $ref in schema follow external $ref and are expanded by spec (e.g. when I call spec.Ref.GetPointer() I never have to iterate)
- $ref in response or parameter outside the schema object (like exposed in this current issue) are only resolved locally at this point and not resolved when calling GetPointer(). It only works when the parameters or responses sections are declared locally in the same document

I don't really know if this behavior is intended or not.

**Comment by kreamyx** [comment]
@fredbi 
```
$ref in response or parameter outside the schema object (like exposed in this current issue) are only resolved locally at this point and not resolved when calling GetPointer(). It only works when the parameters or responses sections are declared locally in the same documentI can't seem to reproduce the second statement
```
I cant reproduce this part. basically all my changes can be seen if you run `swagger flatten` and look at the result
I ran `swagger flatten` and `swagger expand` on the files of this issue and I got the correct flattened/expanded version of the spec. So my guess is that `validate` is probably doing something different?

**Comment by fredbi** [comment]
@kreamyx I don't really understand the subtleties with path setting. What should be done to mimick the flatten behavior?

**Comment by fredbi** [comment]
I did not change validate for this (only shuffled the code) and never really grasped how the ExpandOptions with path should be set. I just let it unchanged. I believe it takes the path from spec.FilePath() or similar

**Comment by kreamyx** [comment]
@fredbi I am not very familiar with the validation code, but my understanding is that validate is merely calling the spec methods to do the expansion? But I would double check that ExpandOptions is populated correctly in the `swagger validate` path.

It might be worth mentioning that I dont recall modifying the Ref.GetPointer() function so if this what is causing an issue then may be the problem is there? (but I might be wrong) 

I can take a look at this later and try to narrow down where the issue exists

**Comment by fredbi** [comment]
in validate all calls to get point have been centralized in helpers.go (because a lot of safeguards are factorized here) and the only explict Expand call is made in spec.go . You may either check yourself or explain how to set the path correctly. This case is illustrated in validate with fixtures/bugs/73

**Comment by kreamyx** [comment]
One thing I quickly noticed was that the panic happens in `validateResponseExample` which is strange because the provided spec has no response examples

**Comment by fredbi** [comment]
Example or default are just spec walkers : they explore the spec in search of examples (resp. defaults). This is what caused the panic.

**Comment by kreamyx** [comment]
I see.
```go
func (s *SpecValidator) validateResponseExample(path string, r *spec.Response) *Result {
	// values provided as example in responses must validate the schema they examplify
	res := new(Result)

	// Recursively follow possible $ref's
	if r.Ref.String() != "" {
		nr, _, err := r.Ref.GetPointer().Get(s.spec.Spec())
		if err != nil {
			s.addPointerError(res, err, r.Ref.String(), strings.Join([]string{"\"" + path + "\"", r.ResponseProps.Schema.ID}, "."))
			return res
		}
		// Here we may expect type assertion to be guaranteed (not like in the Parameter case)
		rr := nr.(spec.Response)
		return s.validateResponseExample(path, &rr)
	}
```
In the validateResponseExample(), is the purpose of the above function to expand the reponse?

**Comment by fredbi** [comment]
you are pointing to the vendored source, right?
there have been much change in go-openapi/validate which has not been vendored yet.
so this code is now somewhere else. but basically, it has not changed much [https://github.com/go-openapi/validate/blob/4d761f0670d764beea9b5495020310edb2bd48f7/helpers.go#L221]

but yes, the intent is to expand the response. I took it in this state and it's still, plus/minus refactoring and safeguarding, doing the same thing

So my problem is not really dealing with the panic (I could grasp that one) but rather
how I could follow external $ref here. 
The test case in current validate master may be run from here: https://github.com/go-openapi/validate/blob/4d761f0670d764beea9b5495020310edb2bd48f7/spec_test.go#L719
_I think we would be better off  reasonning with the current validate master than with the vendored one_

**Comment by kreamyx** [comment]
If yes, then GetPointer() isn't actually the right way to expand the response.
in `spec/expander.go`, there is an unexported method called `expandResponse` that correctly expands responses.
We could have an exported version of this function, something like:
```go
// ExpandResponseWithBasePath expands a response based on a basepath
func ExpandResponseWithBasePath(response *Response, basePath string) error {
	opts := &ExpandOptions {
		RelativeBase: basePath,
	}
	resolver, err := defaultSchemaLoader(nil, opts, nil)
	if err != nil {
		return err
	}

	return expandResponse(response, resolver, basePath)
}
```
And then in `validate/spec.go`, we could replace GetPointer() with
```go
func (s *SpecValidator) validateResponseExample(path string, r *spec.Response) *Result {
	// values provided as example in responses must validate the schema they examplify
	res := new(Result)

	spec.ExpandResponseWithBasePath(r, s.spec.SpecFilePath())

	// NOTE: "examples" in responses vs "example" is a misleading construct in swagger
	if r.Examples != nil {
		if r.Schema != nil {
			if example, ok := r.Examples["application/json"]; ok {
				res.Merge(NewSchemaValidator(r.Schema, s.spec.Spec(), path, s.KnownFormats).Validate(example))
			}

			// TODO: validate other media types too
		}
	}
	return res
}
```
I tested this and the validation passes but I am not sure if there are corner cases I didn't see. If this works, we will need to same thing for parameters as well

**Comment by kreamyx** [comment]
@fredbi , yes I am looking at the vendored version because the issue is reproducible there as well

**Comment by fredbi** [comment]
We will, definitely. That would be a nice thing. 

There is just a small catch here, using some pre-cooked functions from spec.
In validate, I tried to push some evolutions to catch ALL errors and not just stopping on first problem hit.
I know that spec pkg is not very good at "continuing on errors" (I think I posted an issue on that and I might amend some things). 


_Essentially, this is working the same way in vendored and master versions. But the code has been much shuffled around so it might be hard to follow._

**Comment by fredbi** [comment]
I see a spec.ResolveParameter() exported.

**Comment by fredbi** [comment]
This is the one used by the generator.

**Comment by fredbi** [comment]
generator uses ResolveResponse() too

**Comment by kreamyx** [comment]
yes these functions do not expand (follow the $refs)
```go
// Resolve resolves a reference against basePath and stores the result in target
// Resolve is not in charge of following references, it only resolves ref by following its URL
// if the schema that ref is referring to has more refs in it. Resolve doesn't resolve them
// if basePath is an empty string, ref is resolved against the root schema stored in the schemaLoader struct
func (r *schemaLoader) Resolve(ref *Ref, target interface{}, basePath string) error {
	return r.resolveRef(ref, target, basePath)
}
```

**Comment by kreamyx** [comment]
Following $refs are performed by expandResponse() and expandParameter()

**Comment by fredbi** [comment]
ok. let us export these, then. I'll take care of the upgrade in validate.

**Comment by kreamyx** [comment]
Let me know if you have further questions

**Comment by fredbi** [comment]
ok thanks Karim

Fred

**Comment by BertrandGouny** [comment]
thanks :)
